regression: non-defining opaque type use in defining scope

Here is an example which uses a RPIT on a borrowed type in three different ways:
```rust
use std::io::Write;

struct A(Vec<u8>);

struct B<'a> {
  one: &'a mut A,
  two: &'a mut Vec<u8>,
  three: Vec<u8>,
}

impl<'a> B<'a> {
  fn one(&mut self) -> &mut impl Write { &mut self.one.0 }
  fn two(&mut self) -> &mut impl Write { &mut self.two } // fail
  fn three(&mut self) -> &mut impl Write { &mut self.three }
}

struct C<'a>(B<'a>);

impl<'a> C<'a> {
  fn one(&mut self) -> &mut impl Write { self.0.one() } // fail on beta
  fn two(&mut self) -> &mut impl Write { self.0.two() } // fail on beta
  fn three(&mut self) -> &mut impl Write { self.0.three() } // fail on beta
}
```

IMO this is perfectly reasonable code and ~~there should be no errors~~ (EDIT: see below). On stable (`rustc 1.66.0 (69f9c33d7 2022-12-12)`) we have the following:
```
error[E0700]: hidden type for `impl std::io::Write` captures lifetime that does not appear in bounds
  --> src/main.rs:13:42
   |
11 | impl<'a> B<'a> {
   |      -- hidden type `&'a mut Vec<u8>` captures the lifetime `'a` as defined here
12 |   fn one(&mut self) -> &mut impl Write { &mut self.one.0 }
13 |   fn two(&mut self) -> &mut impl Write { &mut self.two } // fail
   |                                          ^^^^^^^^^^^^^
```

And on beta (`rustc 1.67.0-beta.2 (352eb59a4 2022-12-13)`) and nightly (`rustc 1.68.0-nightly (b70baa4f9 2022-12-14)`) we get some additional errors:
```
error[E0700]: hidden type for `impl std::io::Write` captures lifetime that does not appear in bounds
  --> src/main.rs:13:42
   |
11 | impl<'a> B<'a> {
   |      -- hidden type `&'a mut Vec<u8>` captures the lifetime `'a` as defined here
12 |   fn one(&mut self) -> &mut impl Write { &mut self.one.0 }
13 |   fn two(&mut self) -> &mut impl Write { &mut self.two } // fail
   |                                          ^^^^^^^^^^^^^

error: non-defining opaque type use in defining scope
  --> src/main.rs:20:42
   |
20 |   fn one(&mut self) -> &mut impl Write { self.0.one() } // fail on beta
   |                                          ^^^^^^^^^^^^ lifetime `'a` is part of concrete type but not used in parameter list of the `impl Trait` type alias

error: non-defining opaque type use in defining scope
  --> src/main.rs:21:42
   |
21 |   fn two(&mut self) -> &mut impl Write { self.0.two() } // fail on beta
   |                                          ^^^^^^^^^^^^ lifetime `'a` is part of concrete type but not used in parameter list of the `impl Trait` type alias

error: non-defining opaque type use in defining scope
  --> src/main.rs:22:44
   |
22 |   fn three(&mut self) -> &mut impl Write { self.0.three() } // fail on beta
   |                                            ^^^^^^^^^^^^^^ lifetime `'a` is part of concrete type but not used in parameter list of the `impl Trait` type alias
```

EDIT: On second thought, I guess the first error (on stable) is reasonable, since we are actually returning a `&mut &'a mut Vec<u8>` in that case. If you either use `&mut *self.two` in the body or `&mut (impl Write + 'a)` in the return type then the error goes away. But the second error is still a regression.

### Version it worked on

It most recently worked on: rustc 1.66.0 (69f9c33d7 2022-12-12)

### Version with regression

`rustc --version --verbose`:
```
rustc 1.67.0-beta.2 (352eb59a4 2022-12-13)
binary: rustc
commit-hash: 352eb59a4c33abf739914422f2ad975925750146
commit-date: 2022-12-13
host: x86_64-unknown-linux-gnu
release: 1.67.0-beta.2
LLVM version: 15.0.6
```

@rustbot modify labels: +regression-from-stable-to-beta -regression-untriaged

--- a/compiler/rustc_borrowck/src/region_infer/opaque_types.rs
+++ b/compiler/rustc_borrowck/src/region_infer/opaque_types.rs
@@ -250,7 +250,7 @@ impl<'tcx> InferCtxtExt<'tcx> for InferCtxt<'tcx> {
         }
 
         let definition_ty = instantiated_ty
-            .remap_generic_params_to_declaration_params(opaque_type_key, self.tcx, false, origin)
+            .remap_generic_params_to_declaration_params(opaque_type_key, self.tcx, false)
             .ty;
 
         if !check_opaque_type_parameter_valid(

--- a/compiler/rustc_hir_typeck/src/writeback.rs
+++ b/compiler/rustc_hir_typeck/src/writeback.rs
@@ -564,7 +564,6 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
                 opaque_type_key,
                 self.fcx.infcx.tcx,
                 true,
-                decl.origin,
             );
 
             self.typeck_results.concrete_opaque_types.insert(opaque_type_key.def_id, hidden_type);

--- a/compiler/rustc_middle/src/ty/mod.rs
+++ b/compiler/rustc_middle/src/ty/mod.rs
@@ -28,7 +28,6 @@ use crate::ty::util::Discr;
 pub use adt::*;
 pub use assoc::*;
 pub use generics::*;
-use hir::OpaqueTyOrigin;
 use rustc_ast as ast;
 use rustc_ast::node_id::NodeMap;
 use rustc_attr as attr;
@@ -1316,7 +1315,6 @@ impl<'tcx> OpaqueHiddenType<'tcx> {
         tcx: TyCtxt<'tcx>,
         // typeck errors have subpar spans for opaque types, so delay error reporting until borrowck.
         ignore_errors: bool,
-        origin: OpaqueTyOrigin,
     ) -> Self {
         let OpaqueTypeKey { def_id, substs } = opaque_type_key;
 
@@ -1332,30 +1330,7 @@ impl<'tcx> OpaqueHiddenType<'tcx> {
         // This zip may have several times the same lifetime in `substs` paired with a different
         // lifetime from `id_substs`.  Simply `collect`ing the iterator is the correct behaviour:
         // it will pick the last one, which is the one we introduced in the impl-trait desugaring.
-        let map = substs.iter().zip(id_substs);
-
-        let map: FxHashMap<GenericArg<'tcx>, GenericArg<'tcx>> = match origin {
-            // HACK: The HIR lowering for async fn does not generate
-            // any `+ Captures<'x>` bounds for the `impl Future<...>`, so all async fns with lifetimes
-            // would now fail to compile. We should probably just make hir lowering fill this in properly.
-            OpaqueTyOrigin::AsyncFn(_) => map.collect(),
-            OpaqueTyOrigin::FnReturn(_) | OpaqueTyOrigin::TyAlias => {
-                // Opaque types may only use regions that are bound. So for
-                // ```rust
-                // type Foo<'a, 'b, 'c> = impl Trait<'a> + 'b;
-                // ```
-                // we may not use `'c` in the hidden type.
-                let variances = tcx.variances_of(def_id);
-                debug!(?variances);
-
-                map.filter(|(_, v)| {
-                    let ty::GenericArgKind::Lifetime(lt) = v.unpack() else { return true };
-                    let ty::ReEarlyBound(ebr) = lt.kind() else { bug!() };
-                    variances[ebr.index as usize] == ty::Variance::Invariant
-                })
-                .collect()
-            }
-        };
+        let map = substs.iter().zip(id_substs).collect();
         debug!("map = {:#?}", map);
 
         // Convert the type from the function into a type valid outside

--- a/tests/ui/impl-trait/issues/issue-105826.rs
+++ b/tests/ui/impl-trait/issues/issue-105826.rs
@@ -1,0 +1,39 @@
+// check-pass
+
+use std::io::Write;
+
+struct A(Vec<u8>);
+
+struct B<'a> {
+    one: &'a mut A,
+    two: &'a mut Vec<u8>,
+    three: Vec<u8>,
+}
+
+impl<'a> B<'a> {
+    fn one(&mut self) -> &mut impl Write {
+        &mut self.one.0
+    }
+    fn two(&mut self) -> &mut impl Write {
+        &mut *self.two
+    }
+    fn three(&mut self) -> &mut impl Write {
+        &mut self.three
+    }
+}
+
+struct C<'a>(B<'a>);
+
+impl<'a> C<'a> {
+    fn one(&mut self) -> &mut impl Write {
+        self.0.one()
+    }
+    fn two(&mut self) -> &mut impl Write {
+        self.0.two()
+    }
+    fn three(&mut self) -> &mut impl Write {
+        self.0.three()
+    }
+}
+
+fn main() {}

